### DISCUSSION TITLE
left join on non-unique key values

### DISCUSSION THREAD

**Discussion by manuel2a** [hosseinmoein/DataFrame#259] [Q&A] [answered]
Greetings! We're evaluating DataFrame for including in our wasm analytics component. Nice!

One of our requirements is to join DataFrames (like pandas `merge` or `join`)
It seems to me that in DataFrame, join includes only the first matched key and not subsequent matches of the same key.  
* my code below
    * Expected: three records: Bob/Pinto; Mary/Tesla; Mary/Jeep
    * Actual: two records: Bob/Pinto; Mary/Tesla; Jeep is returned unmatched :thinking:
* in [pandas](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.join.html) (see last example)

What am I missing?

---

```
#include <iostream>
#include <vector>

#include <DataFrame/DataFrame.h>

using namespace std;
using namespace hmdf;

using UIDF = StdDataFrame<unsigned int>;

int main( int ia, char** sa ){
  vector<unsigned int> personIndex = { 0, 1 };
  const vector<int> personKey   = { 0, 1 };
  const vector<string> personName = { "Bob", "Mary" };

  vector<unsigned int> carIndex = { 0, 1, 2 };
  const vector<int> carOwnerKey   = { 0, 1, 1 };
  const vector<string> carName = { "Pinto", "Tesla", "Jeep" };

  UIDF People;
  UIDF Cars;

  People.load_index( move( personIndex ));
  People.load_column( "personName", personName );
  People.load_column( "key", personKey );

  Cars.load_index( move( carIndex ));
  Cars.load_column( "carName", carName );
  Cars.load_column( "key", carOwnerKey );

  People  // returns "Jeep", but unmatched (see below)
    .join_by_column<decltype( Cars ), int, string, int>( Cars, "key", join_policy::right_join )
    .write<ostream, string, int, unsigned int>( cout, io_format::csv2 );

  return 0;
}
```

```
0,0,0,0,Bob,Pinto
1,1,1,1,Mary,Tesla
2,0,2,1,,Jeep
```

**Comment by hosseinmoein** [comment, marked as answer]
Thanks for reporting this. This bug was introduced recently.
It is fixed in master now.

**Comment by manuel2a** [comment]
Works! Thank you @hosseinmoein 

```
0,0,0,0,Bob,Pinto
1,1,1,1,Mary,Tesla
2,1,2,1,Mary,Jeep
```

**Comment by hosseinmoein** [comment]
@manuel2a 
If you are comparing DataFrame with Pandas or other data frames. I appreciate it, if you tabulate the results here too.
I always wanted to do that, but never had the time and resources to do it.
